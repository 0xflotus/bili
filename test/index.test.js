import path from 'path'
import Bili from '../src'

function fixture(name) {
  return path.join(__dirname, 'fixtures', name)
}

process.env.BILI_TEST = true

function snapshot({ title, input, ...args }) {
  input = Array.isArray(input) ? input : [input]
  test(title, async () => {
    const { bundles } = await Bili.generate({
      input: input.map(v => fixture(v)),
      ...args
    })
    expect(Object.keys(bundles).map(filepath => [
      bundles[filepath].relative,
      bundles[filepath].code
    ])).toMatchSnapshot()
  })
}

snapshot({
  title: 'defaults',
  input: 'defaults/index.js'
})

snapshot({
  title: 'buble:async',
  input: 'buble/async.js'
})

snapshot({
  title: 'buble:async-and-object-rest-spread',
  input: 'buble/async-dot-dot-dot.js'
})

snapshot({
  title: 'buble:react-jsx',
  input: 'buble/react-jsx.js'
})

snapshot({
  title: 'buble:vue-jsx',
  input: 'buble/vue-jsx.js',
  jsx: 'vue'
})

snapshot({
  title: 'banner:true',
  input: 'banner/index.js',
  banner: true,
  cwd: fixture('banner')
})

snapshot({
  title: 'banner:object',
  input: 'default.js',
  banner: {
    year: '2018',
    author: 'author',
    license: 'GPL',
    name: 'name',
    version: '1.2.3'
  }
})

snapshot({
  title: 'no-js-transform',
  input: 'no-js-transform/index.js',
  js: false
})

snapshot({
  title: 'banner:string',
  input: 'default.js',
  banner: 'woot'
})

snapshot({
  title: 'exclude file',
  input: 'exclude-file/index.js',
  external: ['./test/fixtures/exclude-file/foo.js']
})

snapshot({
  title: 'extendOptions',
  format: 'umd,umd-min,cjs',
  input: ['extend-options/foo.js', 'extend-options/bar.js'],
  extendOptions(options, { input, format, compress }) {
    if (input.endsWith('foo.js')) {
      options.moduleName = 'endsWithFoo'
    }
    if (format === 'umd') {
      options.moduleName = 'umd'
    }
    if (compress) {
      options.moduleName = 'min'
    }
    return options
  }
})

describe('multi formats without suffix error', () => {
  test('it throws', async () => {
    expect.assertions(1)
    try {
      await Bili.generate({
        input: fixture('defaults/index.js'),
        format: ['cjs', 'umd'],
        filename: '[name].js'
      })
    } catch (err) {
      expect(err.message).toMatch(/Multiple files are emitting to the same path/)
    }
  })

  test('it does not throw', async () => {
    await Bili.generate({
      input: fixture('defaults/index.js'),
      format: ['umd-min', 'umd'],
      filename: '[name].js'
    })
  })
})

test('cwd', async () => {
  // This tests two things
  const bili = await Bili.generate({
    // 1. resolve `input` from `cwd`
    input: 'index.js',
    cwd: fixture('defaults')
  })
  // 2. output file is relative to `cwd`
  const outputPath = Object.keys(bili.bundles)[0]
  expect(outputPath).toMatch('/test/fixtures/defaults/dist/index.cjs.js')
})
