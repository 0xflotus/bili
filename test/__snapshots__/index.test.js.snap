// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`banner:object 1`] = `
Array [
  Array [
    "dist/default.cjs.js",
    "/*!
 * name v1.2.3
 * (c) 2018-present author
 * Released under the GPL License.
 */
'use strict';

var _default = {};

module.exports = _default;
",
  ],
]
`;

exports[`banner:string 1`] = `
Array [
  Array [
    "dist/default.cjs.js",
    "woot
'use strict';

var _default = {};

module.exports = _default;
",
  ],
]
`;

exports[`banner:true 1`] = `
Array [
  Array [
    "test/fixtures/banner/dist/index.cjs.js",
    "/*!
 * foo v0.0.0
 * (c) 2015-present name <email@email.com>
 * Released under the MIT License.
 */
'use strict';

var index = 42;

module.exports = index;
",
  ],
]
`;

exports[`buble:async 1`] = `
Array [
  Array [
    "dist/async.cjs.js",
    "'use strict';

var this$1 = undefined;
var a = function () { return new Promise(function ($return, $error) {
  return $return({ a: 'a' });
}.bind(this$1)); };

var async = a().then(function (res) { return Object.assign({}, res); });

module.exports = async;
",
  ],
]
`;

exports[`buble:async-and-object-rest-spread 1`] = `
Array [
  Array [
    "dist/async-dot-dot-dot.cjs.js",
    "'use strict';

var asyncDotDotDot = (function () {
  return new Promise(function ($return, $error) {
    return this.bar().then(function ($await_1) {
      try {
        return $return(Object.assign({}, $await_1));
      } catch ($boundEx) {
        return $error($boundEx);
      }
    }.bind(this), $error);
  }.bind(this));
});

module.exports = asyncDotDotDot;
",
  ],
]
`;

exports[`buble:react-jsx 1`] = `
Array [
  Array [
    "dist/react-jsx.cjs.js",
    "'use strict';

var reactJsx = React.createElement(
  \\"div\\",
  { a: 1 },
  \\"hi\\"
);

module.exports = reactJsx;
",
  ],
]
`;

exports[`buble:vue-jsx 1`] = `
Array [
  Array [
    "dist/vue-jsx.cjs.js",
    "'use strict';

var vueJsx = {
  render: function render() {
    var h = arguments[0];

    return h(
      \\"div\\",
      {
        attrs: { a: 1 }
      },
      [\\"hi\\"]
    );
  }
};

module.exports = vueJsx;
",
  ],
]
`;

exports[`defaults 1`] = `
Array [
  Array [
    "dist/index.cjs.js",
    "'use strict';

var index = 42;

module.exports = index;
",
  ],
]
`;

exports[`exclude file 1`] = `
Array [
  Array [
    "dist/index.cjs.js",
    "'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var foo = _interopDefault(require('./foo.js'));



module.exports = foo;
",
  ],
]
`;

exports[`extendOptions 1`] = `
Array [
  Array [
    "dist/foo.cjs.js",
    "'use strict';

var foo = 'foo';

module.exports = foo;
",
  ],
  Array [
    "dist/bar.cjs.js",
    "'use strict';

var bar = 'bar';

module.exports = bar;
",
  ],
  Array [
    "dist/foo.js",
    "(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.umd = factory());
}(this, (function () { 'use strict';

var foo = 'foo';

return foo;

})));
",
  ],
  Array [
    "dist/bar.js",
    "(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.umd = factory());
}(this, (function () { 'use strict';

var bar = 'bar';

return bar;

})));
",
  ],
  Array [
    "dist/foo.min.js",
    "!function(e,n){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?module.exports=n():\\"function\\"==typeof define&&define.amd?define(n):e.min=n()}(this,function(){\\"use strict\\";return\\"foo\\"});
",
  ],
  Array [
    "dist/bar.min.js",
    "!function(e,n){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?module.exports=n():\\"function\\"==typeof define&&define.amd?define(n):e.min=n()}(this,function(){\\"use strict\\";return\\"bar\\"});
",
  ],
]
`;

exports[`inline:true 1`] = `
Array [
  Array [
    "dist/index.cjs.js",
    "'use strict';

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var pMap = createCommonjsModule(function (module) {
module.exports = (iterable, mapper, opts) => new Promise((resolve, reject) => {
	opts = Object.assign({
		concurrency: Infinity
	}, opts);

	if (typeof mapper !== 'function') {
		throw new TypeError('Mapper function is required');
	}

	const concurrency = opts.concurrency;

	if (!(typeof concurrency === 'number' && concurrency >= 1)) {
		throw new TypeError(\`Expected \\\\\`concurrency\\\\\` to be a number from 1 and up, got \\\\\`\${concurrency}\\\\\` (\${typeof concurrency})\`);
	}

	const ret = [];
	const iterator = iterable[Symbol.iterator]();
	let isRejected = false;
	let iterableDone = false;
	let resolvingCount = 0;
	let currentIdx = 0;

	const next = () => {
		if (isRejected) {
			return;
		}

		const nextItem = iterator.next();
		const i = currentIdx;
		currentIdx++;

		if (nextItem.done) {
			iterableDone = true;

			if (resolvingCount === 0) {
				resolve(ret);
			}

			return;
		}

		resolvingCount++;

		Promise.resolve(nextItem.value)
			.then(el => mapper(el, i))
			.then(
				val => {
					ret[i] = val;
					resolvingCount--;
					next();
				},
				err => {
					isRejected = true;
					reject(err);
				}
			);
	};

	for (let i = 0; i < concurrency; i++) {
		next();

		if (iterableDone) {
			break;
		}
	}
});
});

var pAll = (iterable, opts) => pMap(iterable, el => el(), opts);

var index = { all: pAll, map: pMap };

module.exports = index;
",
  ],
]
`;

exports[`no-js-transform 1`] = `
Array [
  Array [
    "dist/index.cjs.js",
    "'use strict';

var index = () => {}

module.exports = index;
",
  ],
]
`;
